File: callgraph/classify.py
"""Heuristics that map source constructs to layers, external calls, async, and batch.

Edit these tables to match your stack (e.g. KakaoPay's in-house HTTP client or
gRPC stub packages can be added to ``EXTERNAL_SIMPLE_TYPES`` / ``EXTERNAL_PREFIXES``).
"""
from __future__ import annotations

from .model import Layer

# Class-level annotations -> architectural layer.
LAYER_ANNOTATIONS: dict[str, Layer] = {
    "RestController": Layer.CONTROLLER,
    "Controller": Layer.CONTROLLER,
    "Service": Layer.SERVICE,
    "Repository": Layer.REPOSITORY,
    "Component": Layer.COMPONENT,
    "Configuration": Layer.CONFIG,
}

# Spring Data base interfaces: a user interface extending one of these is a repository.
REPOSITORY_BASE_TYPES: set[str] = {
    "Repository",
    "CrudRepository",
    "PagingAndSortingRepository",
    "ListCrudRepository",
    "JpaRepository",
    "ReactiveCrudRepository",
    "CoroutineCrudRepository",
}

# Spring Data methods inherited by repository interfaces (not declared in source).
REPOSITORY_INHERITED_METHODS: set[str] = {
    "save", "saveAll", "saveAndFlush", "findById", "findAll", "findAllById",
    "getById", "getReferenceById", "existsById", "count", "delete", "deleteById",
    "deleteAll", "deleteAllById", "flush",
}

# Simple type names treated as external clients when used as a call receiver.
EXTERNAL_SIMPLE_TYPES: set[str] = {
    "RestTemplate", "WebClient", "RestClient", "JdbcTemplate",
    "NamedParameterJdbcTemplate", "KafkaTemplate", "WebTestClient",
}

# Internal-name prefixes (by import FQCN) treated as external calls.
EXTERNAL_PREFIXES: tuple[str, ...] = (
    "org.springframework.web.client.RestTemplate",
    "org.springframework.web.reactive.function.client.WebClient",
    "org.springframework.web.client.RestClient",
    "org.springframework.jdbc.core.JdbcTemplate",
    "org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate",
    "org.springframework.kafka.core.KafkaTemplate",
    "feign.",
    "org.springframework.cloud.openfeign.",
)

# Annotation that marks a Feign client interface -> every method call on it is external.
FEIGN_CLIENT_ANNOTATIONS: set[str] = {"FeignClient"}

# ---- async detection -------------------------------------------------------

# Method-level annotations that make a method asynchronous.
ASYNC_ANNOTATIONS: set[str] = {"Async", "Scheduled"}

# Kotlin coroutine builders / CompletableFuture builders that open an async context.
ASYNC_BUILDERS: set[str] = {
    "launch", "async", "withContext", "runBlocking", "supervisorScope",
    "coroutineScope", "flow", "produce",
    "supplyAsync", "runAsync", "thenApplyAsync", "thenAcceptAsync", "thenRunAsync",
}

# Return types that imply an asynchronous/reactive method.
ASYNC_RETURN_TYPES: set[str] = {
    "Mono", "Flux", "CompletableFuture", "Future", "Deferred", "Flow", "Publisher",
}

# ---- Spring Batch ----------------------------------------------------------

BATCH_ENABLE_ANNOTATIONS: set[str] = {"EnableBatchProcessing"}

# Base types / annotations marking a class as a batch component.
BATCH_COMPONENT_SUPERTYPES: set[str] = {
    "Tasklet", "ItemReader", "ItemProcessor", "ItemWriter",
    "ItemStreamReader", "ItemStreamWriter", "StepExecutionListener",
    "JobExecutionListener", "ChunkListener",
}

# Builder methods inside a @Bean that wire batch structure, mapped to a relation tag.
# argument identifier -> referenced bean (another @Bean fun in the same config).
BATCH_WIRING_METHODS: dict[str, str] = {
    "start": "batch:step",
    "next": "batch:step",
    "flow": "batch:step",
    "tasklet": "batch:tasklet",
    "reader": "batch:reader",
    "processor": "batch:processor",
    "writer": "batch:writer",
    "listener": "batch:listener",
}

# Keywords that look like calls (``if (...)``) but are not method invocations.
CALL_KEYWORD_BLACKLIST: set[str] = {
    "if", "while", "for", "when", "catch", "switch", "return", "synchronized",
    "fun", "class", "object", "interface", "super", "this", "do", "else",
    "throw", "try", "is", "as", "in", "by", "where", "init", "constructor",
}

File: src/main/kotlin/com/flowmap/callgraph/Classify.kt
package com.flowmap.callgraph

/**
 * Signature tables ported from the Python tool's classify.py.
 * Edit these to match your stack (in-house HTTP/gRPC client packages, etc.).
 */
object Classify {

    val LAYER_ANNOTATIONS: Map<String, Layer> = mapOf(
        "RestController" to Layer.CONTROLLER,
        "Controller" to Layer.CONTROLLER,
        "Service" to Layer.SERVICE,
        "Repository" to Layer.REPOSITORY,
        "Component" to Layer.COMPONENT,
        "Configuration" to Layer.CONFIG,
    )

    val REPOSITORY_BASE_TYPES: Set<String> = setOf(
        "Repository", "CrudRepository", "PagingAndSortingRepository",
        "ListCrudRepository", "JpaRepository", "ReactiveCrudRepository",
        "CoroutineCrudRepository",
    )

    val REPOSITORY_INHERITED_METHODS: Set<String> = setOf(
        "save", "saveAll", "saveAndFlush", "findById", "findAll", "findAllById",
        "getById", "getReferenceById", "existsById", "count", "delete", "deleteById",
        "deleteAll", "deleteAllById", "flush",
    )

    val EXTERNAL_SIMPLE_TYPES: Set<String> = setOf(
        "RestTemplate", "WebClient", "RestClient", "JdbcTemplate",
        "NamedParameterJdbcTemplate", "KafkaTemplate", "WebTestClient",
    )

    val EXTERNAL_PREFIXES: List<String> = listOf(
        "org.springframework.web.client.RestTemplate",
        "org.springframework.web.reactive.function.client.WebClient",
        "org.springframework.web.client.RestClient",
        "org.springframework.jdbc.core.JdbcTemplate",
        "org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate",
        "org.springframework.kafka.core.KafkaTemplate",
        "feign.",
        "org.springframework.cloud.openfeign.",
    )

    // ---- infra resources (Kafka / Redis / DB) ----

    /** Kafka producer: a `send`/`sendDefault` call on one of these receiver types. */
    val KAFKA_TEMPLATE_TYPES: Set<String> = setOf("KafkaTemplate", "ReplyingKafkaTemplate")
    val KAFKA_SEND_METHODS: Set<String> = setOf("send", "sendDefault", "sendOffsetsToTransaction")

    /** Kafka consumer: method annotated with one of these (topics taken from `@KafkaListener`). */
    val KAFKA_LISTENER_ANNOTATIONS: Set<String> = setOf("KafkaListener", "KafkaHandler", "RetryableTopic")

    /** Redis: a call on one of these receiver types -> shared `redis` resource node. */
    val REDIS_TEMPLATE_TYPES: Set<String> = setOf(
        "RedisTemplate", "StringRedisTemplate", "ReactiveRedisTemplate", "ReactiveStringRedisTemplate",
    )

    /** JdbcTemplate-family -> DB resource node (not a 3rd-party HTTP call). */
    val JDBC_TEMPLATE_TYPES: Set<String> = setOf("JdbcTemplate", "NamedParameterJdbcTemplate")

    /** JPA entity marker. */
    val ENTITY_ANNOTATIONS: Set<String> = setOf("Entity")

    /** Spring Data base interfaces whose first type argument is the managed entity. */
    val REPOSITORY_GENERIC_BASES: Set<String> = setOf(
        "JpaRepository", "CrudRepository", "PagingAndSortingRepository", "ListCrudRepository",
        "ReactiveCrudRepository", "CoroutineCrudRepository", "Repository",
    )

    val FEIGN_CLIENT_ANNOTATIONS: Set<String> = setOf("FeignClient")

    /** Spring 6 declarative HTTP interface marker (class- or method-level). */
    val HTTP_EXCHANGE_ANNOTATIONS: Set<String> = setOf("HttpExchange")

    /** Method-level @*Exchange annotations -> HTTP verb. */
    val EXCHANGE_VERBS: Map<String, String> = mapOf(
        "GetExchange" to "GET", "PostExchange" to "POST", "PutExchange" to "PUT",
        "DeleteExchange" to "DELETE", "PatchExchange" to "PATCH",
        "HttpExchange" to "ANY",
    )

    /** Method-level @*Mapping annotations -> HTTP verb. */
    val MAPPING_VERBS: Map<String, String> = mapOf(
        "GetMapping" to "GET", "PostMapping" to "POST", "PutMapping" to "PUT",
        "DeleteMapping" to "DELETE", "PatchMapping" to "PATCH",
        "RequestMapping" to "ANY",
    )

    val ASYNC_ANNOTATIONS: Set<String> = setOf("Async", "Scheduled")

    val ASYNC_BUILDERS: Set<String> = setOf(
        "launch", "async", "withContext", "runBlocking", "supervisorScope",
        "coroutineScope", "flow", "produce",
        "supplyAsync", "runAsync", "thenApplyAsync", "thenAcceptAsync", "thenRunAsync",
    )

    val ASYNC_RETURN_TYPES: Set<String> = setOf(
        "Mono", "Flux", "CompletableFuture", "Future", "Deferred", "Flow", "Publisher",
    )

    val BATCH_ENABLE_ANNOTATIONS: Set<String> = setOf("EnableBatchProcessing")

    val BATCH_COMPONENT_SUPERTYPES: Set<String> = setOf(
        "Tasklet", "ItemReader", "ItemProcessor", "ItemWriter",
        "ItemStreamReader", "ItemStreamWriter", "StepExecutionListener",
        "JobExecutionListener", "ChunkListener",
    )

    val BATCH_RETURN_TYPES: Set<String> = setOf(
        "Job", "Step", "Flow", "Tasklet", "ItemReader", "ItemProcessor", "ItemWriter",
    )

    /** Builder method -> batch relation tag. */
    val BATCH_WIRING_METHODS: Map<String, String> = mapOf(
        "start" to "batch:step",
        "next" to "batch:step",
        "flow" to "batch:step",
        "tasklet" to "batch:tasklet",
        "reader" to "batch:reader",
        "processor" to "batch:processor",
        "writer" to "batch:writer",
        "listener" to "batch:listener",
    )

    fun isExternalByPrefix(fqcn: String?): Boolean =
        fqcn != null && EXTERNAL_PREFIXES.any { fqcn.startsWith(it) }
}
